File: Sources/ImageHub/Views/DashboardView.swift
import SwiftUI

struct DashboardView: View {
    @EnvironmentObject private var appState: AppState
    @Environment(\.appTheme) private var theme

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 18) {
                hero
                checklist
                if !appState.history.isEmpty {
                    recentBuilds
                }
                stats
            }
            .padding(.horizontal, 22)
            .padding(.bottom, 26)
        }
        .safeAreaInset(edge: .top, spacing: 0) {
            PageHeader(
                title: "Overview",
                subtitle: "Wipe a machine, build the drive, reimage from a template."
            ) {
                Button {
                    appState.startBuild(template: nil)
                } label: {
                    Label("Build USB Drive", systemImage: "externaldrive.badge.plus")
                }
                .buttonStyle(.borderedProminent)
                .disabled(appState.isBuilding || appState.templates.templates.isEmpty)
            }
            .background(.bar)
        }
    }

    // MARK: - Hero

    private var hero: some View {
        HStack(spacing: 18) {
            theme.glyph(size: 62)
            VStack(alignment: .leading, spacing: 6) {
                Text(appState.readiness.isReady ? "Ready to build" : "A couple of things to set up")
                    .font(.title3.weight(.semibold))
                Text(
                    appState.readiness.isReady
                        ? "Pick a template, choose the drive, and ImageHub does the rest — wipe, write, and stage the provisioning payload."
                        : "Work through the checklist below and the build button lights up."
                )
                .font(.callout)
                .foregroundStyle(.secondary)
                .fixedSize(horizontal: false, vertical: true)
            }
            Spacer(minLength: 0)
        }
        .glassCard(padding: 20)
    }

    // MARK: - Checklist

    private var checklist: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("Getting ready")
                .font(.headline)

            VStack(spacing: 0) {
                checklistRow(
                    done: appState.readiness.hasTemplate,
                    symbol: "square.stack.3d.up",
                    title: "A buildable deployment template",
                    detail: appState.readiness.hasTemplate
                        ? "\(appState.templates.templates.filter { $0.isBuildable }.count) of \(appState.templates.templates.count) templates are complete."
                        : "Templates need at least a name and an admin password.",
                    actionTitle: "Templates"
                ) {
                    appState.section = .templates
                }

                Divider()

                checklistRow(
                    done: appState.readiness.hasImage,
                    symbol: "opticaldiscdrive",
                    title: "A Windows image in the library",
                    detail: appState.readiness.hasImage
                        ? "\(appState.library.images.filter { $0.fileExists }.count) image(s), \(appState.library.totalBytesOnDisk.byteSize) on disk."
                        : "Download the latest ISO from Microsoft, or import one you already have.",
                    actionTitle: "Images"
                ) {
                    appState.section = .images
                }

                Divider()

                checklistRow(
                    done: appState.readiness.hasDrive,
                    symbol: "externaldrive",
                    title: "A USB drive plugged in",
                    detail: appState.readiness.hasDrive
                        ? appState.drives.map { $0.displayName }.joined(separator: ", ")
                        : "Plug in a 16 GB or larger stick. Internal disks are never offered.",
                    actionTitle: "Drives"
                ) {
                    appState.section = .drives
                }

                if appState.readiness.needsSplitTool {
                    Divider()
                    checklistRow(
                        done: appState.readiness.hasSplitTool,
                        symbol: "scissors",
                        title: "wimlib, for splitting install.wim",
                        detail: appState.readiness.hasSplitTool
                            ? "Found at \(WimTools.locate() ?? "—")."
                            : "Your image's install.wim is over 4 GB, so it must be split to fit the FAT32 boot volume.",
                        actionTitle: "Set up"
                    ) {
                        NSApp.sendAction(
                            Selector(("showSettingsWindow:")), to: nil, from: nil
                        )
                    }
                }
            }
            .glassCard(padding: 0)
        }
    }

    private func checklistRow(
        done: Bool,
        symbol: String,
        title: String,
        detail: String,
        actionTitle: String,
        action: @escaping () -> Void
    ) -> some View {
        HStack(spacing: 12) {
            Image(systemName: done ? "checkmark.circle.fill" : "circle.dashed")
                .font(.system(size: 17))
                .foregroundStyle(done ? AnyShapeStyle(Color.green) : AnyShapeStyle(.tertiary))
                .frame(width: 22)

            Image(systemName: symbol)
                .font(.system(size: 13))
                .foregroundStyle(.secondary)
                .frame(width: 18)

            VStack(alignment: .leading, spacing: 2) {
                Text(title)
                    .font(.callout.weight(.medium))
                Text(detail)
                    .font(.caption)
                    .foregroundStyle(.secondary)
                    .lineLimit(2)
                    .fixedSize(horizontal: false, vertical: true)
            }

            Spacer(minLength: 8)

            Button(actionTitle, action: action)
                .buttonStyle(.bordered)
                .controlSize(.small)
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 12)
    }

    // MARK: - Recent builds

    private var recentBuilds: some View {
        VStack(alignment: .leading, spacing: 12) {
            HStack {
                Text("Recent builds")
                    .font(.headline)
                Spacer()
                Button("See all") { appState.section = .builds }
                    .buttonStyle(.link)
                    .font(.callout)
            }

            VStack(spacing: 0) {
                ForEach(Array(appState.history.prefix(3))) { job in
                    BuildSummaryRow(job: job)
                    if job.id != appState.history.prefix(3).last?.id {
                        Divider()
                    }
                }
            }
            .glassCard(padding: 0)
        }
    }

    // MARK: - Stats

    private var stats: some View {
        HStack(spacing: 12) {
            statCard(
                value: "\(appState.templates.templates.count)",
                label: "Templates",
                symbol: "square.stack.3d.up"
            )
            statCard(
                value: "\(appState.library.images.count)",
                label: "Images",
                symbol: "opticaldiscdrive"
            )
            statCard(
                value: appState.library.totalBytesOnDisk.byteSize,
                label: "On disk",
                symbol: "internaldrive"
            )
            statCard(
                value: "\(appState.drives.count)",
                label: "USB drives",
                symbol: "externaldrive"
            )
        }
    }

    private func statCard(value: String, label: String, symbol: String) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            Image(systemName: symbol)
                .font(.system(size: 15))
                .foregroundStyle(theme.primary)
            Text(value)
                .font(.title3.weight(.semibold))
                .lineLimit(1)
                .minimumScaleFactor(0.7)
            Text(label)
                .font(.caption)
                .foregroundStyle(.secondary)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .glassCard(padding: 14)
    }
}

/// One row in a build list: state, names, elapsed time.
struct BuildSummaryRow: View {
    @ObservedObject var job: BuildJob

    var body: some View {
        HStack(spacing: 12) {
            Image(systemName: symbol)
                .font(.system(size: 15))
                .foregroundStyle(tint)
                .frame(width: 20)

            VStack(alignment: .leading, spacing: 2) {
                Text(job.templateName)
                    .font(.callout.weight(.medium))
                Text("→ \(job.driveName)")
                    .font(.caption)
                    .foregroundStyle(.secondary)
            }

            Spacer(minLength: 8)

            VStack(alignment: .trailing, spacing: 2) {
                Text(stateLabel)
                    .font(.caption.weight(.medium))
                    .foregroundStyle(tint)
                if let elapsed = job.elapsed {
                    Text(Self.duration(elapsed))
                        .font(.caption2)
                        .foregroundStyle(.tertiary)
                }
            }
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 11)
    }

    private var symbol: String {
        switch job.phase {
        case .idle: return "clock"
        case .running: return "arrow.triangle.2.circlepath"
        case .succeeded: return "checkmark.circle.fill"
        case .failed: return "xmark.circle.fill"
        case .cancelled: return "slash.circle"
        }
    }

    private var tint: Color {
        switch job.phase {
        case .succeeded: return .green
        case .failed: return .red
        case .cancelled: return .orange
        default: return .secondary
        }
    }

    private var stateLabel: String {
        switch job.phase {
        case .idle: return "Queued"
        case .running: return "\(Int(job.overallProgress * 100))%"
        case .succeeded: return "Ready"
        case .failed: return "Failed"
        case .cancelled: return "Cancelled"
        }
    }

    static func duration(_ seconds: TimeInterval) -> String {
        let total = Int(seconds)
        if total < 60 { return "\(total)s" }
        let minutes = total / 60
        let remainder = total % 60
        if minutes < 60 { return "\(minutes)m \(remainder)s" }
        return "\(minutes / 60)h \(minutes % 60)m"
    }
}

File: Windows/App/Ui/DashboardView.cs
using System;
using System.Linq;
using System.Windows;
using System.Windows.Controls;
using ImageHub.Models;
using ImageHub.Services;
using ImageHub.Support;
using ImageHub.ViewModels;

namespace ImageHub.Views;

/// <summary>
/// The Overview page: what still has to be set up before a drive can be built, and
/// what has been built recently.
///
/// The checklist has one more item than the macOS app's — administrator rights — and
/// one fewer: nothing to install for splitting an oversized install.wim, because DISM
/// is already here.
/// </summary>
public sealed class DashboardView : UserControl, IRefreshable
{
    private readonly MainWindow _window;
    private readonly AppState _state = AppState.Shared;
    private readonly StackPanel _body = new();

    public DashboardView(MainWindow window)
    {
        _window = window;
        Content = Ui.Page(
            Ui.PageHeader("Overview", "Wipe a machine, build the drive, reimage from a template.",
                Ui.Button("Build USB Drive", () => _window.StartBuild(null), "AccentButton")),
            Ui.Scroll(_body));
        Refresh();
    }

    public void Refresh()
    {
        AppState.ReadinessReport readiness = _state.Readiness;
        _body.Children.Clear();

        _body.Children.Add(Hero(readiness));
        _body.Children.Add(Checklist(readiness));
        if (_state.History.Count > 0) { _body.Children.Add(RecentBuilds()); }
        _body.Children.Add(Stats());
    }

    private UIElement Hero(AppState.ReadinessReport readiness)
    {
        var mark = new System.Windows.Shapes.Path
        {
            Data = Glyphs.Drive,
            Stretch = System.Windows.Media.Stretch.Uniform,
            Width = 44,
            Height = 44,
            StrokeThickness = 1.2,
            VerticalAlignment = VerticalAlignment.Top,
        };
        mark.Themed(System.Windows.Shapes.Path.StrokeProperty, "AccentBrush");

        TextBlock heading = Ui.Heading(readiness.IsReady
            ? "Ready to build"
            : "A couple of things to set up");
        TextBlock body = Ui.Caption(readiness.IsReady
            ? "Pick a template, choose the drive, and ImageHub does the rest — wipe, write, and "
              + "stage the provisioning payload."
            : "Work through the checklist below and the build button lights up.");
        body.MaxWidth = 640;

        StackPanel row = Ui.Row(16, mark, Ui.Column(5, heading, body));
        row.VerticalAlignment = VerticalAlignment.Top;
        Border card = Ui.Card(row, 18);
        card.Margin = new Thickness(0, 0, 0, 6);
        return card;
    }

    private UIElement Checklist(AppState.ReadinessReport readiness)
    {
        int buildable = _state.Templates.Templates.Count(template => template.IsBuildable);
        int usable = _state.Library.Images.Count(image => image.FileExists);

        return Ui.Group("Getting ready",
            ChecklistRow(
                readiness.HasTemplate,
                "A buildable deployment template",
                readiness.HasTemplate
                    ? $"{buildable} of {_state.Templates.Templates.Count} templates are complete."
                    : "Templates need at least a name and an admin password.",
                "Templates",
                () => _state.CurrentSection = Section.Templates),
            ChecklistRow(
                readiness.HasImage,
                "A Windows image in the library",
                readiness.HasImage
                    ? $"{Formatting.Plural(usable, "image")}, "
                      + $"{Formatting.ByteSize(_state.Library.TotalBytesOnDisk)} on disk."
                    : "Get the ISO from Microsoft in your browser, then import it.",
                "Images",
                () => _state.CurrentSection = Section.Images),
            ChecklistRow(
                readiness.HasDrive,
                "A USB drive plugged in",
                readiness.HasDrive
                    ? string.Join(", ", _state.Drives.Select(drive => drive.DisplayName))
                    : "Plug in a 16 GB or larger stick. Internal disks are never offered.",
                "Drives",
                () => _state.CurrentSection = Section.Drives),
            ChecklistRow(
                readiness.IsElevated,
                "Administrator rights",
                readiness.IsElevated
                    ? "ImageHub is running elevated, so it can erase a drive and mount an ISO."
                    : "Erasing a drive and mounting an ISO both need them. Everything else doesn't, "
                      + "which is why ImageHub doesn't ask at launch.",
                readiness.IsElevated ? null : "Restart as Administrator",
                () =>
                {
                    if (Elevation.RelaunchElevated()) { Application.Current.Shutdown(); }
                },
                shield: true));
    }

    private UIElement ChecklistRow(
        bool done,
        string title,
        string detail,
        string? actionTitle,
        Action action,
        bool shield = false)
    {
        var grid = new Grid();
        grid.ColumnDefinitions.Add(new ColumnDefinition { Width = GridLength.Auto });
        grid.ColumnDefinitions.Add(new ColumnDefinition { Width = new GridLength(1, GridUnitType.Star) });
        grid.ColumnDefinitions.Add(new ColumnDefinition { Width = GridLength.Auto });

        System.Windows.Shapes.Path tick = Ui.Icon(
            done ? Glyphs.Success : Glyphs.Circle, 18,
            done ? "SuccessBrush" : "TextTertiary", thickness: 1.5);
        tick.VerticalAlignment = VerticalAlignment.Center;
        tick.Margin = new Thickness(0, 0, 12, 0);
        Grid.SetColumn(tick, 0);
        grid.Children.Add(tick);

        StackPanel text = Ui.Column(2, Ui.Text(title), Ui.Caption(detail));
        text.VerticalAlignment = VerticalAlignment.Center;
        text.Margin = new Thickness(0, 0, 14, 0);
        Grid.SetColumn(text, 1);
        grid.Children.Add(text);

        if (actionTitle is not null)
        {
            Button button = shield
                ? Ui.ShieldButton(actionTitle, action)
                : Ui.Button(actionTitle, action);
            button.VerticalAlignment = VerticalAlignment.Center;
            Grid.SetColumn(button, 2);
            grid.Children.Add(button);
        }

        Border card = Ui.Card(grid, 0);
        card.Padding = new Thickness(14, 12, 14, 12);
        card.Margin = new Thickness(0, 0, 0, 6);
        return card;
    }

    private UIElement RecentBuilds()
    {
        var rows = new System.Collections.Generic.List<UIElement?>();
        foreach (BuildJob job in _state.History.Take(3))
        {
            rows.Add(BuildRow.Card(job, () => _state.CurrentSection = Section.Builds));
        }
        rows.Add(Ui.Button("See all builds", () => _state.CurrentSection = Section.Builds, "LinkButton"));
        return Ui.Group("Recent builds", rows.ToArray());
    }

    private UIElement Stats()
    {
        var grid = new Grid { Margin = new Thickness(0, 4, 0, 0) };
        for (int i = 0; i < 4; i++)
        {
            grid.ColumnDefinitions.Add(new ColumnDefinition { Width = new GridLength(1, GridUnitType.Star) });
        }

        (string Value, string Label)[] stats =
        {
            (_state.Templates.Templates.Count.ToString(), "Templates"),
            (_state.Library.Images.Count.ToString(), "Images"),
            (Formatting.ByteSize(_state.Library.TotalBytesOnDisk), "On disk"),
            (_state.Drives.Count.ToString(), "USB drives"),
        };

        for (int i = 0; i < stats.Length; i++)
        {
            var value = new TextBlock
            {
                Text = stats[i].Value,
                FontSize = 19,
                FontWeight = FontWeights.SemiBold,
            };
            value.Themed(TextBlock.ForegroundProperty, "TextPrimary");
            Border card = Ui.Card(Ui.Column(3, value, Ui.Caption(stats[i].Label)), 14);
            card.Margin = new Thickness(i == 0 ? 0 : 3, 0, i == stats.Length - 1 ? 0 : 3, 0);
            Grid.SetColumn(card, i);
            grid.Children.Add(card);
        }
        return grid;
    }
}

/// <summary>One row describing a build: state, names, elapsed time. Shared by two pages.</summary>
public static class BuildRow
{
    public static Border Card(BuildJob job, Action? click = null)
    {
        var grid = new Grid();
        grid.ColumnDefinitions.Add(new ColumnDefinition { Width = GridLength.Auto });
        grid.ColumnDefinitions.Add(new ColumnDefinition { Width = new GridLength(1, GridUnitType.Star) });
        grid.ColumnDefinitions.Add(new ColumnDefinition { Width = GridLength.Auto });

        System.Windows.Shapes.Path icon = Ui.Icon(GlyphFor(job), 17, BrushFor(job), thickness: 1.5);
        icon.VerticalAlignment = VerticalAlignment.Center;
        icon.Margin = new Thickness(0, 0, 12, 0);
        Grid.SetColumn(icon, 0);
        grid.Children.Add(icon);

        StackPanel text = Ui.Column(2,
            Ui.Text(job.TemplateName),
            Ui.Caption("→ " + job.DriveName));
        text.VerticalAlignment = VerticalAlignment.Center;
        Grid.SetColumn(text, 1);
        grid.Children.Add(text);

        var state = new TextBlock
        {
            Text = StateLabel(job),
            FontSize = 12,
            FontWeight = FontWeights.Medium,
            HorizontalAlignment = HorizontalAlignment.Right,
        };
        state.Themed(TextBlock.ForegroundProperty, BrushFor(job));
        StackPanel right = Ui.Column(2, state,
            job.Elapsed is null
                ? null
                : Ui.Hint(Formatting.ShortDuration(job.Elapsed.Value)));
        right.VerticalAlignment = VerticalAlignment.Center;
        right.HorizontalAlignment = HorizontalAlignment.Right;
        Grid.SetColumn(right, 2);
        grid.Children.Add(right);

        Border card = Ui.Card(grid, 0);
        card.Padding = new Thickness(14, 11, 14, 11);
        card.Margin = new Thickness(0, 0, 0, 6);
        if (click is not null)
        {
            card.Cursor = System.Windows.Input.Cursors.Hand;
            card.MouseLeftButtonUp += (_, _) => click();
        }
        return card;
    }

    public static System.Windows.Media.Geometry GlyphFor(BuildJob job) => job.CurrentPhase switch
    {
        BuildJob.Phase.Succeeded => Glyphs.Success,
        BuildJob.Phase.Failed => Glyphs.Error,
        BuildJob.Phase.Cancelled => Glyphs.Minus,
        _ => Glyphs.Circle,
    };

    public static string BrushFor(BuildJob job) => job.CurrentPhase switch
    {
        BuildJob.Phase.Succeeded => "SuccessBrush",
        BuildJob.Phase.Failed => "DangerBrush",
        BuildJob.Phase.Cancelled => "WarningBrush",
        _ => "TextSecondary",
    };

    public static string StateLabel(BuildJob job) => job.CurrentPhase switch
    {
        BuildJob.Phase.Idle => "Queued",
        BuildJob.Phase.Running => $"{(int)(job.OverallProgress * 100)}%",
        BuildJob.Phase.Succeeded => "Ready",
        BuildJob.Phase.Failed => "Failed",
        BuildJob.Phase.Cancelled => "Cancelled",
        _ => string.Empty,
    };
}
